<a href="https://colab.research.google.com/github/MariaDaguer/Analitica-de-negocios/blob/main/CarCrash_Reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

CAR CRASH

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
nxl="/content/1. BD2_CarCrash_Int.xlsx"
XDB=pd.read_excel(nxl, sheet_name=1)

#Variables
XDB=XDB[['Weekday','ViolCat','ClearWeather','Month','Crash Type','Highway','Daylight']]


XD=XDB[['Weekday','ViolCat','ClearWeather','Month','Crash Type','Daylight']]
yd=XDB[['Highway']]

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd)


ct=mnb.class_count_
print("Como están separados los datos",ct) #accidentes en una autopista o no
print("El total de los datos es:",len(XD))
print("Los datos porcentualmente están divididos en (No accidentes- accidentes en la autopista):", ct/len(XD)) # % de accidentes en autopista o no


# Se obtiene la información de los accidentes por categoría
#las categorías son accidenten es la autopista o no en la autopista
u=mnb.theta_
print("La media de los accidentes por categoría es:\n",u)

std=mnb.var_ # esto es varianza
std=np.sqrt(std) #esto es desviación estándar
print("La desviación estándar de los accidentes por categoría es:\n",std)

# valores superiores e inferiores de los accidentes
print("El limite superior de los valores:\n", u+std)
print("El limite inferior de los valores:\n", u-std)




Como están separados los datos [80539. 32121.]
El total de los datos es: 112660
Los datos porcentualmente están divididos en (No accidentes- accidentes en la autopista): [0.7148855 0.2851145]
La media de los accidentes por categoría es:
 [[3.90307801 6.79464607 0.89238754 6.67500217 3.80866413 0.69460758]
 [3.92883161 4.45913888 0.86345382 6.68867719 3.49905047 0.69487251]]
La desviación estándar de los accidentes por categoría es:
 [[1.93312956 3.60184338 0.30989035 3.36048976 1.49341633 0.46057345]
 [1.91313258 2.47514729 0.34336764 3.33713602 1.19979219 0.46046142]]
El limite superior de los valores:
 [[ 5.83620757 10.39648945  1.20227789 10.03549193  5.30208046  1.15518103]
 [ 5.84196418  6.93428617  1.20682146 10.02581321  4.69884266  1.15533393]]
El limite inferior de los valores:
 [[1.96994846 3.1928027  0.58249719 3.31451241 2.31524779 0.23403413]
 [2.01569903 1.98399159 0.52008618 3.35154117 2.29925827 0.2344111 ]]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)
predecir= mnb.predict(XS)
print(predecir)

[0]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [ ]:
ydp=mnb.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VN+VP)/(VN+FP+FN+VP)
print("La exactitud es:",Ex)


TE= (FN+VP)/(VN+FP+FN+VP)
print("La tasa de error es:",TE)


Sen=VP/(VP+FN)
print("La sensibilidad es:",Sen)


Esp=VN/(VN+FP)
print("La especificidad es:",Esp)


Pre=VP/(VP+FP)
print("La precisión es:",Pre)

Pren=VN/(VN+FN)
print("La predicción negativa es:",Pren)


[[69410 11129]
 [21498 10623]]
La exactitud es: 0.7103941061601278
La tasa de error es: 0.2851145038167939
La sensibilidad es: 0.3307182217241057
La especificidad es: 0.8618184978705968
La precisión es: 0.48836888561971314
La predicción negativa es: 0.7635191622299468


Análisis

EL ejercicio de este análisis es predecir si un accidente automovilistico se produce o no en una autopista, para esto se consideraron variables como el día, mes, condiciones climáticas y lúminicas, tipo de accidente e impacto. Estas variables se condensaron en una base de datos con un total de 112.660 datos disponibles, que al ser categorizados se dividieron en un 28% para accidentes ocurridos en la autopista y un 72% para los que no, esto se traduce a un desequilibrio en los datos que puede conllevar a mayores imprecisiones en las predicciones; Por ejemplo, al observar la media de las distintas variables, se ven los valores superiores en la categoría de no accidentes, especialmente en variables como el nivel de impacto y el tipo de accidente.

Al ejecutar este modelo, pudimos observar que tiene una baja sensibilidad (33%) y una alta especificidad(86%) lo que indica que el modelo es mejor al predecir los casos de no accidentes pero tiene dificultades para predecir accidentes. Cabe aclarar que modelo de Bayes posee una exactitud que supera el 70%, convirtiendolo por tanto, en una fuente ligeramente imprecisa pero aun así capáz.